coilsnake/modules/eb/StaffModule.py

```diff
@@ -0,0 +1,268 @@
+                        ENTRY_TYPE, ENTRY_TILE, vtile, ENTRY_CHAR, vchar
+                    )
+                )
+
+    def read_staff_chars_from_assets(self):
+        with open_asset('structures', 'eb_staff_chars.yml') as f:
+            yml_data = yml_load(f)
+
+        self.read_staff_chars(yml_data)
+
+    def read_staff_chars_from_project(self, resource_open):
+        with resource_open(STAFF_CHARS_FILE_NAME, 'yml', True) as f:
+            yml_data = yml_load(f)
+
+        self.read_staff_chars(yml_data)
+
+    def read_text_line_from_project(self, byte, line, mapping, height):
+        self.data.append(byte)
+
+        if line == '':
+            line = ' '
+
+        for char in line:
+            self.data.append(mapping[char])
+
+        self.data.append(CONTROL_END_OF_LINE)
+        self.height += height
+
+    def read_small_line_from_project(self, line):
+        self.read_text_line_from_project(CONTROL_SMALL, line, self.small, HEIGHT_SMALL)
+
+    def read_big_line_from_project(self, line):
+        self.read_text_line_from_project(CONTROL_BIG, line, self.big, HEIGHT_BIG)
+
+    def read_space_from_project(self, how_much):
+        if how_much == 0: # Ignore zero-sized space
+            return
+
+        self.data.append(CONTROL_SPACE)
+        self.data.append(how_much)
+        self.height += HEIGHT_SMALL * how_much
+
+    def read_keyword_from_project(self, keyword):
+        if keyword not in KEYWORD_BYTE_HEIGHT:
+            raise InvalidUserDataError(
+                'Invalid keyword={} in {}.yml - should be one of {}.'.format(
+                    keyword, name, KEYWORD_BYTE_HEIGHT.keys()
+                )
+            )
+
+        byte, height = KEYWORD_BYTE_HEIGHT[keyword]
+        self.data.append(byte)
+        self.height += height
+
+    def read_staff_text_from_project(self, resource_open):
+        with resource_open(STAFF_TEXT_FILE_NAME, 'md', True) as f:
+            line = f.readline()
+
+            while line:
+                if line[-1] == '\n':
+                    line = line[:-1]
+
+                line.strip()
+
+                if line == '':
+                    line = f.readline()
+                    continue
+
+                text = line[1:].lstrip()
+                mark = line[0]
+                line = f.readline()
+
+                if mark == '>':
+                    self.read_space_from_project(int(text))
+                elif mark not in ['#', '-']:
+                    self.read_keyword_from_project(mark + text)
+                elif len(text) > SCREEN_WIDTH_IN_TILES:
+                    raise InvalidUserDataError(
+                        'Text \'{}\' is too long - must have at most {} characters'.format(
+                            text, SCREEN_WIDTH_IN_TILES
+                        )
+                    )
+                elif mark == '#':
+                    self.read_small_line_from_project(text)
+                else:
```

Comment:
what if the first character is some invalid character? people often make mistakes with these kinds of files, so I think it'd be good to have an error message for that instead of just assuming it's a `-` (if I'm reading this code right

Reply:
This was tested in the line `elif mark not in ['#', '-']` and then `read_keyword_from_project` will complain that it's not a keyword.

Reply:
oh right, my mistake